### ISSUE TITLE
Uptime Calculation using Push Monitor is Wrong

### ISSUE THREAD

**Issue by filipre** [louislam/uptime-kuma#919]
### ⚠️ Please verify that this bug has NOT been raised before.

- [X] I checked and didn't find similar issue

### 🛡️ Security Policy

- [X] I agree to have read this project [Security Policy](https://github.com/louislam/uptime-kuma/security/policy)

### 📝 Describe your problem

Hi!

I am monitoring a service using the push monitor. Today, my service was down for 20 seconds and the uptime percentage dropped to less than 50%. After a few hours, it "recovered" to only 55%. I am wondering, how does the percentage gets calculated? I'd have expect to see it recover to 99% or something.

I attached a 24h log and also the events:

```
Up | 2021-11-14 18:44:20 | OK
Down | 2021-11-14 18:44:02 | No heartbeat in the time window
Up | 2021-11-14 00:34:11 | OK
```

<img width="921" alt="Screenshot 2021-11-14 at 23 28 19" src="https://user-images.githubusercontent.com/8502198/141701271-c56729df-cfe9-405a-be5f-ce281dcd216c.png">

Is there something I am missing?

### 🐻 Uptime-Kuma Version

1.10.1

### 💻 Operating System and Arch

Raspberian

### 🌐 Browser

Chrome

### 🐋 Docker Version

Docker version 20.10.8, build 3967b7d

**Comment by DatDraggy** [comment]
Hi, I believe I'm having the same issue. My router restarted, so a few services had a downtime for a few minutes. The "Home Assistant" ones % went down by less than one. It's a web monitor.

All the others are Push monitors and are down by much more than expected. Those are python sending heartbeats

![image](https://user-images.githubusercontent.com/11386054/153273893-5e033371-fcc6-4561-ba25-3567cd2a1f8d.png)

Same with heartbeats sent from bash/curl

![image](https://user-images.githubusercontent.com/11386054/153274443-2582e5b5-2a89-4752-ae7a-d9fe3d176d16.png)

**Comment by marvinscham** [comment]
Same issue here, instance using Docker 20.10.12 on Raspbian Buster, UK 1.11.4

**Comment by filipre** [comment]
Let's get to the bottom of the problem. We know that the [Uptime.vue](https://github.com/louislam/uptime-kuma/blob/master/src/components/Uptime.vue) component shows the wrong number. It uses the uptime information coming from `$root.uptimeList[key]` where `let key = this.monitor.id + "_" + this.type;` and it is being populated by the [api-router.js](https://github.com/louislam/uptime-kuma/blob/master/server/routers/api-router.js):
```js
uptimeList[`${monitorID}_${type}`] = await Monitor.calcUptime(type, monitorID);
```
This leads us to  [monitor.js](https://github.com/louislam/uptime-kuma/blob/master/server/model/monitor.js) being responsible for the uptime calculation. Here, `type` here refers to the duration? The final uptime is then calculated by
```js
uptime = uptimeDuration / totalDuration;
```
Since the uptime calculation works for normal monitors, I expect that we do not write correct rows into the `heartbeat` table when we use a Push Monitor. The heatbeat is calculated in the [monitor.js](https://github.com/louislam/uptime-kuma/blob/master/server/model/monitor.js#L288). I noticed that all other types set `bean.status = UP;` but not the push monitor. Could this be somehow related? Or something to do with this?
```js
// No need to insert successful heartbeat for push type, so end here
```
Does someone have ideas how to debug this?

**Comment by karelkryda** [comment]
+1 for this issue

My monitor has never been offline (I only see one line in the monitor history with the text "UP") and the percentage is calculated randomly (sometimes 100%, sometimes around 50%, ...).

![image](https://user-images.githubusercontent.com/59953812/167375130-90bbbff8-2392-4ea9-8ba0-eb5eddf3eba4.png)

**Comment by deluxghost** [comment]
> +1 for this issue
> 
> My monitor has never been offline (I only see one line in the monitor history with the text "UP") and the percentage is calculated randomly (sometimes 100%, sometimes around 50%, ...).
> 
> ![image](https://user-images.githubusercontent.com/59953812/167375130-90bbbff8-2392-4ea9-8ba0-eb5eddf3eba4.png)
> 

sometimes, it seems like the uptime is based on the current time in a day (increase from 0% to 100%), but sometimes it looks like more randomly than that

**Comment by karelkryda** [comment]
Does the problem also occur outside of Docker and Raspbian OS?

**Comment by karelkryda** [comment]
> Let's get to the bottom of the problem. We know that the [Uptime.vue](https://github.com/louislam/uptime-kuma/blob/master/src/components/Uptime.vue) component shows the wrong number. It uses the uptime information coming from `$root.uptimeList[key]` where `let key = this.monitor.id + "_" + this.type;` and it is being populated by the [api-router.js](https://github.com/louislam/uptime-kuma/blob/master/server/routers/api-router.js):
> 
> ```js
> uptimeList[`${monitorID}_${type}`] = await Monitor.calcUptime(type, monitorID);
> ```
> 
> This leads us to [monitor.js](https://github.com/louislam/uptime-kuma/blob/master/server/model/monitor.js) being responsible for the uptime calculation. Here, `type` here refers to the duration? The final uptime is then calculated by
> 
> ```js
> uptime = uptimeDuration / totalDuration;
> ```
> 
> Since the uptime calculation works for normal monitors, I expect that we do not write correct rows into the `heartbeat` table when we use a Push Monitor. The heatbeat is calculated in the [monitor.js](https://github.com/louislam/uptime-kuma/blob/master/server/model/monitor.js#L288). I noticed that all other types set `bean.status = UP;` but not the push monitor. Could this be somehow related? Or something to do with this?
> 
> ```js
> // No need to insert successful heartbeat for push type, so end here
> ```
> 
> Does someone have ideas how to debug this?

in the current version, the data from the push monitor is stored in the database in the same way as the data from other monitors.

![image](https://user-images.githubusercontent.com/59953812/167387062-ea817c15-f1b3-413b-aff4-20e5396c5be6.png)
*id 1 - Ping monitor*
*id 2 - Push monitor*

all code logic looks the same for push monitor and other types of monitors.

I'm currently testing the raw source code on Windows 11. Could there be a problem only in the Docker environment on RPi (Raspbian OS)?



I will try to get the database file from my Docker instance and see.

**Comment by karelkryda** [comment]
**got it**

![image](https://user-images.githubusercontent.com/59953812/167390055-7b92c34c-6dca-4e38-8480-7e9e312047c5.png)
![image](https://user-images.githubusercontent.com/59953812/167390093-92a7d547-8c35-4dcd-97f8-a86ad3e63e08.png)
![image](https://user-images.githubusercontent.com/59953812/167390146-1119dd1e-e7d9-4eed-a63c-f66eb15d3bc1.png)
![image](https://user-images.githubusercontent.com/59953812/167392835-2aa7f709-5ff8-43c6-8448-90ee30297d15.png)

it looks like every night (last few records around 03:14) an unreasonably large duration value is stored.

I pulled out a table where the duration exceeds 1000 and returned me to basically all push monitors.

PS: a few records can of course be real outages (but there has been no real outage for at least a week).
![image](https://user-images.githubusercontent.com/59953812/167390777-4988538c-cac0-4a8b-9f84-ce6a89761716.png)


## DB pulled from
Docker version: **20.10.13, build a224086**
Uptime Kuma version: **1.15.1**
OS: **Raspbian GNU/Linux 10 (buster)**

**Comment by karelkryda** [comment]
OK, so the problem must be here:
https://github.com/louislam/uptime-kuma/blob/5bf58cc6c4430226cc558deb0cc1738e6899baee/server/model/monitor.js#L204

This file is the only one in which, in the case of a push monitor, status 2 (PENDING) is stored in the database, and this is the only place where the duration is calculated.

\
In my experiments, the calculation on this line should work normally. I simulated the process using a playground, see. photo below.
I entered the time when the hearthbeat (PENDING) was recorded and the time of the previous hearthbeat and the duration value corresponds to expected value (187 seconds).
![image](https://user-images.githubusercontent.com/59953812/167404813-ad90a3eb-16d2-4130-89b4-7857a739f9df.png)

\
Also the previous heartbeat from the database should be correct (really the last available).

So it looks like one of these problems:
  - bad `previousBeat` from database
  - error in `dayjs.diff()` function
  - incompatibility with Docker / Raspbian?
